File: packages/core/src/sync/save-paths.ts
import { basename, join } from "node:path";
import type { IndexedRomFile } from "../db/index.js";
import { isRetroArchSyncPlatform } from "../platform-emulator-map.js";

/** RetroArch battery-save extensions probed under `{saves_path}/{esde_folder}/`. */
export const BATTERY_SAVE_EXTENSIONS = [
  ".srm",
  ".sav",
  ".rtc",
  ".eep",
  ".fla",
  ".mcr",
  ".mcd",
  ".vmp",
  ".cds",
  ".bkr",
  ".bcr",
  ".smpc",
] as const;

/** RetroArch save-state suffixes under `{states_path}/{esde_folder}/`. */
export const STATE_FILE_SUFFIXES = [
  ".state",
  ".state0",
  ".state1",
  ".state2",
  ".state3",
  ".state4",
  ".state5",
  ".state6",
  ".state7",
  ".state8",
  ".state9",
] as const;

export type SaveFileKind = "battery" | "state";

export interface ExpectedSavePath {
  rom_id: number;
  esde_folder: string;
  absolutePath: string;
  file_name: string;
  kind: SaveFileKind;
  slot: string;
  emulator: "retroarch";
}

/** ROM filename without extension; preserves `(USA)` and other tags. */
export function romBasename(filename: string): string {
  const base = basename(filename);
  const dot = base.lastIndexOf(".");
  if (dot <= 0) return base;
  return base.slice(0, dot);
}

/** RomM slotted uploads embed a UTC timestamp tag before the extension. */
const DATETIME_TAG_PATTERN = / \[\d{4}-\d{2}-\d{2}_\d{2}-\d{2}-\d{2}\]/;

/** Strip RomM datetime tags so paths match RetroArch filenames. */
export function untagSaveFileName(fileName: string): string {
  return fileName.replace(DATETIME_TAG_PATTERN, "");
}

/** Extension (including dot) from a server or local save basename. */
export function saveFileExtension(fileName: string): string {
  const untagged = untagSaveFileName(fileName);
  const dot = untagged.lastIndexOf(".");
  if (dot <= 0) return "";
  return untagged.slice(dot);
}

/** Map a RomM save filename to the local RetroArch basename for an indexed ROM. */
export function resolveLocalSaveFileName(
  indexedRomFilename: string,
  serverFileName: string,
): string {
  const ext = saveFileExtension(serverFileName);
  if (!ext) return untagSaveFileName(serverFileName);
  return `${romBasename(indexedRomFilename)}${ext}`;
}

/** RomM pairs saves on `(rom_id, slot)` — each state slot needs a distinct slot name. */
export function slotForSaveFileName(fileName: string): string {
  const lower = fileName.toLowerCase();
  const match = lower.match(/\.state(\d*)$/);
  if (match) {
    const suffix = match[1];
    return suffix === "" ? "state" : `state${suffix}`;
  }
  return "default";
}

export function isStateFileName(fileName: string): boolean {
  return /\.state\d*$/i.test(fileName);
}

export function resolveExpectedSavePaths(row: IndexedRomFile, roots: {
  savesPath: string;
  statesPath: string;
}): ExpectedSavePath[] {
  if (!isRetroArchSyncPlatform(row.esde_folder)) return [];

  const base = romBasename(row.filename);
  const out: ExpectedSavePath[] = [];

  for (const ext of BATTERY_SAVE_EXTENSIONS) {
    const file_name = `${base}${ext}`;
    out.push({
      rom_id: row.rom_id,
      esde_folder: row.esde_folder,
      absolutePath: join(roots.savesPath, row.esde_folder, file_name),
      file_name,
      kind: "battery",
      slot: "default",
      emulator: "retroarch",
    });
  }

  for (const suffix of STATE_FILE_SUFFIXES) {
    const file_name = `${base}${suffix}`;
    out.push({
      rom_id: row.rom_id,
      esde_folder: row.esde_folder,
      absolutePath: join(roots.statesPath, row.esde_folder, file_name),
      file_name,
      kind: "state",
      slot: slotForSaveFileName(file_name),
      emulator: "retroarch",
    });
  }

  return out;
}

/** Canonical local path for a downloaded or uploaded save/state file. */
export function resolveLocalSavePath(
  roots: { savesPath: string; statesPath: string },
  esdeFolder: string,
  fileName: string,
): string {
  const root = isStateFileName(fileName) ? roots.statesPath : roots.savesPath;
  return join(root, esdeFolder, fileName);
}

/** Dedupe indexed rows by `(rom_id, esde_folder, filename)`. */
export function uniqueIndexedRomFiles(rows: IndexedRomFile[]): IndexedRomFile[] {
  const seen = new Set<string>();
  const out: IndexedRomFile[] = [];
  for (const row of rows) {
    const key = `${row.rom_id}\0${row.esde_folder}\0${row.filename}`;
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(row);
  }
  return out;
}

File: src/shared/src/commonMain/kotlin/dev/rommdeck/shared/sync/SavePaths.kt
package dev.rommdeck.shared.sync

import dev.rommdeck.shared.db.IndexedRomFile
import dev.rommdeck.shared.io.joinPath
import dev.rommdeck.shared.io.readClasspathResource
import kotlinx.serialization.json.Json

private val json = Json { ignoreUnknownKeys = true }

private val emulatorMap: Map<String, String> by lazy {
    val text = readClasspathResource("platform-emulator-map.json") ?: return@lazy emptyMap()
    json.decodeFromString<Map<String, String>>(text)
}

fun isRetroArchSyncPlatform(esdeFolder: String): Boolean {
    val family = emulatorMap[esdeFolder] ?: return true
    return family == "retroarch"
}

val BATTERY_SAVE_EXTENSIONS = listOf(
    ".srm", ".sav", ".rtc", ".eep", ".fla", ".mcr", ".mcd",
    ".vmp", ".cds", ".bkr", ".bcr", ".smpc",
)

val STATE_FILE_SUFFIXES = listOf(
    ".state", ".state0", ".state1", ".state2", ".state3",
    ".state4", ".state5", ".state6", ".state7", ".state8", ".state9",
)

enum class SaveFileKind { BATTERY, STATE }

data class ExpectedSavePath(
    val romId: Int,
    val esdeFolder: String,
    val absolutePath: String,
    val fileName: String,
    val kind: SaveFileKind,
    val slot: String,
    val emulator: String = "retroarch",
)

fun romBasename(filename: String): String {
    val base = filename.substringAfterLast('/').substringAfterLast('\\')
    val dot = base.lastIndexOf('.')
    if (dot <= 0) return base
    return base.substring(0, dot)
}

private val DATETIME_TAG = Regex(" \\[\\d{4}-\\d{2}-\\d{2}_\\d{2}-\\d{2}-\\d{2}\\]")

fun untagSaveFileName(fileName: String): String = fileName.replace(DATETIME_TAG, "")

fun saveFileExtension(fileName: String): String {
    val untagged = untagSaveFileName(fileName)
    val dot = untagged.lastIndexOf('.')
    if (dot <= 0) return ""
    return untagged.substring(dot)
}

fun resolveLocalSaveFileName(indexedRomFilename: String, serverFileName: String): String {
    val ext = saveFileExtension(serverFileName)
    if (ext.isEmpty()) return untagSaveFileName(serverFileName)
    return "${romBasename(indexedRomFilename)}$ext"
}

fun slotForSaveFileName(fileName: String): String {
    val lower = fileName.lowercase()
    val match = Regex("\\.state(\\d*)$").find(lower) ?: return "default"
    val suffix = match.groupValues[1]
    return if (suffix.isEmpty()) "state" else "state$suffix"
}

fun isStateFileName(fileName: String): Boolean = Regex("\\.state\\d*$", RegexOption.IGNORE_CASE).containsMatchIn(fileName)

fun resolveExpectedSavePaths(
    row: IndexedRomFile,
    savesPath: String,
    statesPath: String,
): List<ExpectedSavePath> {
    if (!isRetroArchSyncPlatform(row.esdeFolder)) return emptyList()
    val base = romBasename(row.filename)
    val out = mutableListOf<ExpectedSavePath>()
    for (ext in BATTERY_SAVE_EXTENSIONS) {
        val fileName = "$base$ext"
        out += ExpectedSavePath(
            romId = row.romId,
            esdeFolder = row.esdeFolder,
            absolutePath = joinPath(savesPath, row.esdeFolder, fileName),
            fileName = fileName,
            kind = SaveFileKind.BATTERY,
            slot = "default",
        )
    }
    for (suffix in STATE_FILE_SUFFIXES) {
        val fileName = "$base$suffix"
        out += ExpectedSavePath(
            romId = row.romId,
            esdeFolder = row.esdeFolder,
            absolutePath = joinPath(statesPath, row.esdeFolder, fileName),
            fileName = fileName,
            kind = SaveFileKind.STATE,
            slot = slotForSaveFileName(fileName),
        )
    }
    return out
}

fun resolveLocalSavePath(
    savesPath: String,
    statesPath: String,
    esdeFolder: String,
    fileName: String,
): String {
    val root = if (isStateFileName(fileName)) statesPath else savesPath
    return joinPath(root, esdeFolder, fileName)
}

fun uniqueIndexedRomFiles(rows: List<IndexedRomFile>): List<IndexedRomFile> {
    val seen = mutableSetOf<String>()
    val out = mutableListOf<IndexedRomFile>()
    for (row in rows) {
        val key = "${row.romId}\u0000${row.esdeFolder}\u0000${row.filename}"
        if (key in seen) continue
        seen += key
        out += row
    }
    return out
}
